probe_src/probe_frontend/src/main.rs

```diff
@@ -0,0 +1,325 @@
+mod display;
+
+/// Parsing of arena directories created by libprobe into a cross-platform
+/// serialized format.
+///
+/// # Serialization format
+///
+/// The serialization format output is very similar to the raw libprobe arena format. It's a
+/// filesystem hierarchy of `<PID>/<EXEC_EPOCH>/<TID>` but instead of `<TID>` being a directory containing
+/// `ops` and `data` directories with the raw C-struct arenas, `<TID>` is a
+/// [jsonlines](https://jsonlines.org/) file, where each line is a json representation of an
+/// [`ops::Op`].
+mod arena;
+
+/// System metadata recorded into probe logs.
+mod metadata;
+
+
+
+/// Generate or manipulate Provenance for Replay OBservation Engine (PROBE) logs.
+#[derive(clap::Parser, Debug, Clone)]
+#[command(author, version, about, long_about = None)]
+#[command(propagate_version = true)]
+struct Cli {
+    #[command(subcommand)]
+    command: Command,
+}
+
+#[derive(clap::Subcommand, Debug, Clone)]
+enum Command {
+    /// Execute a command and record its provenance
+    Record {
+        /// Directory to output PROBE log to
+        #[arg(short, long, required = false, default_value = "probe_log")]
+        output: OsString,
+
+        /// Overwrite existing output directory if it exists
+        #[arg(short = 'f', long)]
+        overwrite: bool,
+
+        /// Run in gdb
+        #[arg(long)]
+        gdb: bool,
+
+        /// Override the path to libprobe.so (this path will be canonicalized)
+        #[arg(long)]
+        lib_path: Option<PathBuf>,
+
+        /// Run in verbose & debug build of libprobe
+        #[arg(long)]
+        debug: bool,
+
+        /// Command to execute under provenance
+        #[arg(required = true)]
+        cmd: Vec<OsString>,
+    },
+
+    /// Write the data from probe log data in a human-readable manner
+    Dump {
+        /// Directory to load PROBE log from
+        #[arg(short, long, required = false, default_value = "probe_log")]
+        input: OsString,
+    },
+}
+
+// TODO: break out each sub-command as a separate function
+fn main() -> Result<()> {
+    color_eyre::install()?;
+    env_logger::Builder::from_env(env_logger::Env::new().filter_or("__PROBE_LOG", "warn")).init();
+    log::info!("Logger Facility Initialized");
+
+    match Cli::parse().command {
+        Command::Record {
+            output,
+            overwrite,
+            gdb,
+            lib_path,
+            debug,
+            cmd,
+        } => {
+            // if -f is set, we should clear-out the old probe_log
+            if overwrite {
+                match fs::remove_file(&output) {
+                    Ok(_) => (),
+                    Err(e) => match e.kind() {
+                        std::io::ErrorKind::NotFound => (),
+                        _ => return Err(e).wrap_err("Error deleting old output file"),
+                    },
+                };
+            }
+
+            // the path to the libprobe.so directory is searched for as follows:
```

Comment:
I think we should always "bundle" libprobe with this wrapper. If we search some paths, there could be a mismatch between the version of libprobe and the version of the wrapper. I also think search paths where they aren't necessary, should be avoided; they smell too much like FHS and unpredictability to me.

I think the best way to "operationalize" the bundling is to load from a single location relative to the wrapper, which can be a symlink. In Git, the symlink will point to the current build; in Nix derivation, we can change the symlink to wherever the library is.

Alternatively, we could do the same thing with env var and `wrapProgram`. Let me know your thoughts.

Reply:
That all makes sense, one consideration I was trying to keep in mind was "How would this work on a FHS system" where presumably `libprobe.so` would be in a `share` folder of some kind (I assumed not `lib` because It was my understanding that it's not designed for direct linking), while the binary would be in a `bin` folder, you could of course install to `/opt` and then symlink to a `bin`, but that's hacky at best.

Obviously I don't have a magic solution to this problem, if I did, I would've written a paper on my perfect solution to dependency management.

Reply:
Addressed in https://github.com/charmoniumQ/PROBE/pull/20/commits/a83cce757de26a1bf559602fcd74f7f2b6df6705; now loads only from environment variable set by the devShell (and eventually a wrapper).